WebApplication/Migration.cs

```diff
@@ -0,0 +1,164 @@
+// DOES NOT WARRANT THAT THE OPERATION OF THE PROGRAM WILL BE
+// UNINTERRUPTED OR ERROR FREE.
+/////////////////////////////////////////////////////////////////////
+
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Autodesk.Forge.Model;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using WebApplication.Definitions;
+using WebApplication.Processing;
+using WebApplication.Services;
+using WebApplication.State;
+using WebApplication.Utilities;
+using static MigrationApp.MigrationJob;
+
+namespace MigrationApp
+{
+   public class Migration
+   {
+      private readonly IConfiguration _configuration;
+      private readonly BucketPrefixProvider _bucketPrefix;
+      private readonly IForgeOSS _forgeOSS;
+      private readonly MigrationBucketKeyProvider _bucketProvider;
+      private readonly ResourceProvider _resourceProvider;
+      private readonly UserResolver _userResolver;
+      private readonly ProjectWork _projectWork;
+      private readonly ILogger<Migration> _logger;
+
+      public Migration(IConfiguration configuration, BucketPrefixProvider bucketPrefix, IForgeOSS forgeOSS, MigrationBucketKeyProvider bucketProvider, UserResolver userResolver, ProjectWork projectWork, ILogger<Migration> logger, ResourceProvider resourceProvider)
+      {
+         _forgeOSS = forgeOSS;
+         _configuration = configuration;
+         _bucketPrefix = bucketPrefix;
+         _bucketProvider = bucketProvider;
+         _userResolver = userResolver;
+         _projectWork = projectWork;
+         _logger = logger;
+         _resourceProvider = resourceProvider;
+      }
+      public async Task<List<MigrationJob>> ScanBuckets()
+      {
+         List<MigrationJob> migrationJobs = new List<MigrationJob>();
+         string suffixFrom = _configuration.GetValue<string>("BucketKeySuffixOld");
+         string bucketKeyStart = _bucketPrefix.GetBucketPrefix(suffixFrom);
+         List<string> bucketKeys = await _forgeOSS.GetBucketsAsync();
+         foreach (string bucketKey in bucketKeys)
+         {
+            if (bucketKey.Length >= bucketKeyStart.Length && 
+               bucketKey.Substring(0, bucketKeyStart.Length) == bucketKeyStart ||
+               bucketKey == _resourceProvider.AnonymousBucketKey(suffixFrom)
+               )
+            {
+               await ScanBucket(migrationJobs, bucketKey);
+            }
+         }
+
+         return migrationJobs;
+      }
+
+      private async Task ScanBucket(List<MigrationJob> migrationJobs, string bucketKey)
+      {
+         List<ObjectDetails> ossProjectFiles = await _forgeOSS.GetBucketObjectsAsync(bucketKey, "projects/");
+         foreach (ObjectDetails file in ossProjectFiles)
+         {
+            string projectUrl = file.ObjectKey;
+            string projectName = projectUrl.Split('/')[1];
+            string attributeFile = "attributes/" + projectName + "/metadata.json";
```

Comment:
This is a location of metadata file on oss, not in the cache. If you have valid point feel free to commit it.

Reply:
Here is how it could look like:
```csharp
var ossAttributes = new OssAttributes(projectName);
string metadataName = ossAttributes.Metadata;

// check attributes file existence in new destination bucket
string bucketKeyNew = _bucketProvider.SetBucketKeyFromOld(bucketKey);
var newBucket = new OssBucket(_forgeOSS, bucketKeyNew, _logger);
if (await newBucket.ObjectExistsAsync(metadataName)) return;

var oldBucket = new OssBucket(_forgeOSS, bucketKey, _logger);
var projectMetadata = await oldBucket.DeserializeAsync<ProjectMetadata>(metadataName);

var projectInfo = new ProjectInfo { Name = projectName, TopLevelAssembly = projectMetadata.TLA };

migrationJobs.Add(new MigrationJob(JobType.CopyAndAdopt, bucketKey, projectInfo, projectUrl));
```

As further improvement - `oldBucket` can be passed into `MigrationJob` instead of `bucketKey`, so its methods can be used right away.

Reply:
done